libqtile/layout/wmii.py

```diff
@@ -25,443 +16,34 @@
-        cur = self.current_window
-        if cur is None:
-            return
-        for c in self.columns:
-            if cur in c['rows']:
-                cidx = self.columns.index(c)
-                if self.is_last_column(cidx):
-                    # grow right from right most, shrink right
-                    if cidx == 0:
-                        return
-                    self.swap_column_width(cidx - 1, cidx)
-                    self.group.focus(cur)
-                    return
-                # grow my width by 20, reduce neighbor to the right by 20
-                self.swap_column_width(cidx, cidx + 1)
-                self.group.focus(cur)
-                return
-
-    def cmd_shuffle_right(self):
-        cur = self.current_window
-        if cur is None:
-            return
-        for c in self.columns:
-            if cur in c['rows']:
-                cidx = self.columns.index(c)
-                if self.is_last_column(cidx):
-                    if len(c['rows']) == 1:
-                        return
-                    c['rows'].remove(cur)
-                    self.add_column(False, cur)
-                    if len(c['rows']) == 0:
-                        self.columns.remove(c)
-                else:
-                    c['rows'].remove(cur)
-                    self.columns[cidx + 1]['rows'].append(cur)
-                if len(c['rows']) == 0:
-                    self.columns.remove(c)
-                    newwidth = int(100 / len(self.columns))
-                    for c in self.columns:
-                        c['width'] = newwidth
-                else:
-                    if c['active'] >= len(c['rows']):
-                        c['active'] = len(c['rows']) - 1
-                self.group.focus(cur)
-                return
-
-    def cmd_shuffle_down(self):
-        for c in self.columns:
-            if self.current_window in c['rows']:
-                r = c['rows']
-                ridx = r.index(self.current_window)
-                if ridx + 1 < len(r):
-                    r[ridx], r[ridx + 1] = r[ridx + 1], r[ridx]
-                    client = r[ridx + 1]
-                    self.focus(client)
-                    self.group.focus(client)
-                return
-
-    def cmd_shuffle_up(self):
-        for c in self.columns:
-            if self.current_window in c['rows']:
-                r = c['rows']
-                ridx = r.index(self.current_window)
-                if ridx > 0:
-                    r[ridx - 1], r[ridx] = r[ridx], r[ridx - 1]
-                    client = r[ridx - 1]
-                    self.focus(client)
-                    self.group.focus(client)
-                return
+        warnings.warn(
+            "Wmii layout is deprecated in favor of Columns.",
+            category=DeprecationWarning,
+            stacklevel=2)
+        for key, value, _ in Wmii.defaults:
+            if key not in config:
+                config[key] = value
```

Comment:
Can't we simply do something like this?

```python
Columns.__init__(self, **config)
self.add_defaults(Wmii.defaults)
```

Reply:
That's how it was before. The problem is that on initialization, Columns creates a Column with the current values of `split` and `insert_position`. That first instance of Column will incorrectly be created with Column.defaults instead of Wmii.defaults.

Reply:
Ah, yeah, alright!

Let's just replace the `if` block with `dict.setdefault` then? 😉